src/lsm/cache_map.zig

```diff
@@ -0,0 +1,358 @@
+                // Scope Map: Case 1.
+                self.last_upsert_was_update_with_eviction = true;
+                if (self.scope_is_active) {
+                    _ = self.scope_map.putAssumeCapacity(value.*, {});
+                }
+            } else {
+                if (self.scope_is_active) {
+                    // Scope Map: Case 3.
+                    _ = self.scope_map.putAssumeCapacity(value.*, {});
+
+                    const gop = self.map_1.getOrPutAssumeCapacity(value.*);
+                    gop.key_ptr.* = value.*;
+                } else {
+                    const gop = self.map_1.getOrPutAssumeCapacity(value.*);
+                    gop.key_ptr.* = value.*;
+                }
+            }
+        }
+
+        pub fn remove(self: *Self, key: Key) void {
+            // The only thing that tests this in any depth is the cache_map fuzz itself.
+            // Make sure we aren't being called in regular code without another once over.
+            assert(constants.verify);
+
+            const maybe_removed = self.cache.remove(key);
+
+            if (maybe_removed) |removed| {
+                if (self.scope_is_active) {
+                    _ = self.scope_map.putAssumeCapacity(removed, {});
+                }
+            } else {
+                if (self.scope_is_active) {
+                    // TODO: Actually, does the fuzz catch this...
+                    // TODO: So if we delete from map_2 and put to map_1, there's a problem;
+                    //       because when we undo our scope we insert back to map_1 :/
+                    var maybe_map_removed = self.map_1.getKey(tombstone_from_key(key)) orelse
+                        self.map_2.getKey(tombstone_from_key(key));
+                    if (maybe_map_removed) |map_removed| {
+                        _ = self.scope_map.putAssumeCapacity(map_removed, {});
+                    }
+                }
+            }
+
+            // We always need to try remove from the stash; since it could have a stale value.
+            _ = self.map_1.remove(tombstone_from_key(key));
+            _ = self.map_2.remove(tombstone_from_key(key));
+        }
+
+        /// Start a new scope. Within a scope, changes can be persisted
+        /// or discarded. Only one scope can be active at a time.
+        pub fn scope_open(self: *Self) void {
+            assert(!self.scope_is_active);
+            self.scope_is_active = true;
+        }
+
+        pub fn scope_close(self: *Self, data: ScopeCloseMode) void {
+            assert(self.scope_is_active);
+            self.scope_is_active = false;
+
+            // We don't need to do anything to persist a scope.
+            if (data == .persist) {
+                self.scope_map.clearRetainingCapacity();
+                return;
+            }
+
+            // The scope_map stores the operations we need to reverse the changes a scope made.
+            // Replay them back.
+            var it = self.scope_map.iterator();
```

Comment:
We only need to store / replay the first item, so a map does buy us some absorption. One optimization down the line is we might actually pass our table_mutable in (since it _is_ a log) and not record inserts into the `scope_map`.

Then, the insert case with a scope active is free; rollbacks would walk the table_mutable log and delete entries, before using the scope_map only for reverting updates / removes.